addons/web/static/src/core/macro.js

```diff
@@ -0,0 +1,213 @@
+                if (this.onTimeout) {
+                    const index = this.currentIndex;
+                    const step = this.steps[index];
+                    this.safeCall(this.onTimeout, step, index);
+                } else {
+                    const error = new TimeoutError("Step timeout");
+                    this.handleError(error);
+                }
+            }, this.timeoutDuration);
+        }
+    }
+
+    handleError(error) {
+        // mark the macro as complete, so it can be cleaned up from the
+        // engine
+        this.isComplete = true;
+        clearTimeout(this.timeout);
+        if (this.onError) {
+            const index = this.currentIndex;
+            const step = this.steps[index];
+            this.onError(error, step, index);
+        } else {
+            console.error(error);
+        }
+    }
+}
+
+class MacroEngine {
+    constructor() {
+        this.isRunning = false;
+        this.timeout = null;
+        this.interval = Infinity; // nbr of ms before we check the dom to advance macros
+        this.macros = new Set();
+        this.observer = new MutationObserver(this.delayedCheck.bind(this));
+    }
+
+    async activate(descr) {
+        // micro task tick to make sure we add the macro in a new call stack,
+        // so we are guaranteed that we are not iterating on the current macros
+        await Promise.resolve();
+        const macro = new Macro(descr);
+        this.interval = Math.min(this.interval, macro.interval);
+        this.macros.add(macro);
+        this.start();
+    }
+
+    start() {
+        if (!this.isRunning) {
+            this.isRunning = true;
+            this.observer.observe(document.body, {
+                attributes: true,
+                childList: true,
+                subtree: true,
+                attributeOldValue: true,
+                characterData: true,
+            });
+        }
+        this.delayedCheck();
+    }
+
+    stop() {
+        if (this.isRunning) {
+            this.isRunning = false;
+            clearTimeout(this.timeout);
+            this.timeout = null;
+            this.observer.disconnect();
+        }
+    }
+
+    delayedCheck() {
+        if (this.timeout) {
+            clearTimeout(this.timeout);
+        }
+        this.timeout = setTimeout(this.advanceMacros.bind(this), this.interval);
+    }
+
+    advanceMacros() {
+        let toDelete = [];
+        for (let macro of this.macros) {
+            macro.advance();
+            if (macro.isComplete) {
+                toDelete.push(macro);
+            }
+        }
+        if (toDelete.length) {
+            for (let macro of toDelete) {
+                this.macros.delete(macro);
+            }
+            // recompute current interval, because it may need to be increased
+            this.interval = Infinity;
+            for (let macro of this.macros) {
+                this.interval = Math.min(this.interval, macro.interval);
```

Comment:
Kind of confused on the utility of "interval". Is there any reason we'd like to configure this? If I understand correctly this is the "batch size", ie how long there must be between 2 DOM mutations before we advance the macros, but I'm not sure why we would want to make this configurable per macro.